File: tic-tac-toe-javafx-java/src/main/java/com/mj/tic/tac/toe/javafx/java/controller/ControllerMediator.java
package com.mj.tic.tac.toe.javafx.java.controller;

import javafx.concurrent.Task;

/**
 * Defines the communication contract used by child controllers to delegate
 * background task execution to a coordinating controller.
 *
 * <p>This mediator keeps controllers loosely coupled by preventing child
 * controllers from directly depending on the parent controller's implementation
 * details. A child controller can submit a JavaFX {@link Task} through this
 * interface, while the mediator decides how the task should be configured,
 * scheduled, and observed.</p>
 *
 * <p>Implementations should keep execution details, such as the underlying
 * executor service or task event handling policy, hidden from callers. This
 * allows the application to change its threading strategy without changing each
 * controller that needs to run background work.</p>
 *
 * @author Montaser Jamal
 * @email montaser.jjs@gmail.com
 * @phone +962-786258874
 * @created 29-05-2026
 * @since 1.0.0
 */

public interface ControllerMediator {

    /**
     * Schedules the given JavaFX task for execution.
     *
     * <p>Controllers should call this method when they need to run work in the
     * background and allow the mediator to handle task execution consistently.
     * The mediator implementation may attach common event handlers, submit the
     * task to an executor, or apply any other application-level task management
     * behavior.</p>
     *
     * @param task the JavaFX task to execute.
     */
    void execute(Task<?> task);
}

File: tic-tac-toe-javafx-kotlin/src/main/kotlin/com/mj/tic/tac/toe/javafx/kotlin/controller/ControllerMediator.kt
package com.mj.tic.tac.toe.javafx.kotlin.controller

import javafx.concurrent.Task

/**
 * Functional interface that decouples controller-to-controller
 * communication by abstracting background task execution.
 *
 * Part of the Mediator pattern: child controllers (e.g.,
 * [com.mj.tic.tac.toe.javafx.kotlin.controller.layout.LeftPanelController])
 * do not execute tasks directly but instead delegate to the mediator,
 * which is implemented by [ViewController]. This centralizes task
 * scheduling, result handling, and error management.
 *
 * Defined as a Kotlin [fun interface] (SAM), allowing both
 * class-based and lambda-based implementations.
 *
 * @author Montaser Sbaih
 * @version 1.0
 * @email montaser.jjs@gmail.com
 * @phone +962-786258874
 * @since 29-05-2026
 */

fun interface ControllerMediator {

    /**
     * Executes the given background task.
     *
     * The implementation is responsible for scheduling the task on
     * an appropriate thread (typically a background executor) and
     * attaching result-handling callbacks.
     *
     * @param task The [Task] to execute. The task's result type is
     *   unchecked; the mediator implementation knows which concrete
     *   task types to expect and how to route their results.
     */
    fun execute(task: Task<*>)
}
